File: 187.重复的dna序列.py
#
# @lc app=leetcode.cn id=187 lang=python
#
# [187] 重复的DNA序列
#
# https://leetcode-cn.com/problems/repeated-dna-sequences/description/
#
# algorithms
# Medium (42.02%)
# Likes:    25
# Dislikes: 0
# Total Accepted:    3.1K
# Total Submissions: 7.4K
# Testcase Example:  '"AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"'
#
# 所有 DNA 由一系列缩写为 A，C，G 和 T 的核苷酸组成，例如：“ACGAATTCCG”。在研究 DNA 时，识别 DNA
# 中的重复序列有时会对研究非常有帮助。
# 
# 编写一个函数来查找 DNA 分子中所有出现超多一次的10个字母长的序列（子串）。
# 
# 示例:
# 
# 输入: s = "AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"
# 
# 输出: ["AAAAACCCCC", "CCCCCAAAAA"]
# 
#
class Solution(object):
    def findRepeatedDnaSequences(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        

File: 187.重复的dna序列.java
/*
 * @lc app=leetcode.cn id=187 lang=java
 *
 * [187] 重复的DNA序列
 *
 * https://leetcode-cn.com/problems/repeated-dna-sequences/description/
 *
 * algorithms
 * Medium (48.37%)
 * Likes:    223
 * Dislikes: 0
 * Total Accepted:    47.4K
 * Total Submissions: 96K
 * Testcase Example:  '"AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"'
 *
 * 所有 DNA 都由一系列缩写为 'A'，'C'，'G' 和 'T' 的核苷酸组成，例如："ACGAATTCCG"。在研究 DNA 时，识别 DNA
 * 中的重复序列有时会对研究非常有帮助。
 *
 * 编写一个函数来找出所有目标子串，目标子串的长度为 10，且在 DNA 字符串 s 中出现次数超过一次。
 *
 *
 *
 * 示例 1：
 *
 *
 * 输入：s = "AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"
 * 输出：["AAAAACCCCC","CCCCCAAAAA"]
 *
 *
 * 示例 2：
 *
 *
 * 输入：s = "AAAAAAAAAAAAA"
 * 输出：["AAAAAAAAAA"]
 *
 *
 *
 *
 * 提示：
 *
 *
 * 0
 * s[i] 为 'A'、'C'、'G' 或 'T'
 *
 *
 */
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.ArrayList;

// @lc code=start
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        List<String> ret = new ArrayList<String>();
        Map<String, Integer> freq = new HashMap<String, Integer>();
        for (int i = 0; i <= s.length() - 10; i++) {
            String sub = s.substring(i, i + 10);
            Integer subCnt = freq.getOrDefault(sub, 0);
            freq.put(sub, subCnt + 1);
            if (freq.get(sub) == 2) {
                ret.add(sub);
            }
        }
        return ret;
    }
}
// @lc code=end

